### PR TITLE
Addresses #4647, wrap the E+ Table:Lookup, Table:IndependentVariableList, and Table:IndependentVariable objects

### LINKED ISSUE #4647: Wrap the E+ Table:Lookup, Table:IndependentVariableList, and Table:IndependentVariable objects.
Enhancement Request

Coming from @asparke2 from an email request. 

Wrap the E+ Table:Lookup, Table:IndependentVariableList, and Table:IndependentVariable objects. OS has the previous E+ incarnation of the LookupTable objects wrapped, but I believe those were replaced with these new ones. This new format is what Eric Bonnema used to represent the Carrier HP performance.

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #4647
 - [x] Deprecate `TableMultiVariableLookup`
   - [x] Throw warning from ctor
   - [x] Throw error from FT
   - [x] Replace old with new in ForwardTranslator.cpp, and then throw error in old FT
   - [ ] Support External File?
   - [ ] ~Create VT from old to new (note: this is not required if the previous item is addressed)~
 - [x] Model
   - [x] Source files
   - [x] Tests
 - [x] FT
   - [x] Source files
   - [x] Tests
 - [ ] ~RT?~
   - [ ] ~Source files~
   - [ ] ~Tests~

**Possible implementation:**
  * `Table:Lookup` (Curve?)
  * `Table:IndependentVariableList` (Use `ModelObjectList`?)
  * `Table:IndependentVariable` (ParentObject?)

much like the E+ IDD:
```
  Table:IndependentVariable,
    HPACCoolCapFFF_IndependentVariable1,  !- Name
    Cubic,                   !- Interpolation Method
    Constant,                !- Extrapolation Method
    0.5,                     !- Minimum Value
    1.5,                     !- Maximum Value
    ,                        !- Normalization Reference Value
    Dimensionless,           !- Unit Type
    ,                        !- External File Name
    ,                        !- External File Column Number
    ,                        !- External File Starting Row Number
    0.0,                     !- Value 1
    1.0,                     !- Value 2
    1.5;                     !- <none>

  Table:IndependentVariableList,
    HPACCoolCapFFF_IndependentVariableList,  !- Name
    HPACCoolCapFFF_IndependentVariable1;  !- Independent Variable 1 Name

  Table:Lookup,
    HPACCoolCapFFF,          !- Name
    HPACCoolCapFFF_IndependentVariableList,  !- Independent Variable List Name
    ,                        !- Normalization Method
    ,                        !- Normalization Divisor
    0.8,                     !- Minimum Output
    1.5,                     !- Maximum Output
    Dimensionless,           !- Output Unit Type
    ,                        !- External File Name
    ,                        !- External File Column Number
    ,                        !- External File Starting Row Number
    0.8,                     !- Output Value 1
    1.0,                     !- Output Value 2
    1.1;                     !- <none>
```

### Pull Request Author

 - [x] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by jmarrec** [comment]
@joseph-robertson, some thoughts:


> **Possible implementation:**
> 
>     * `Table:Lookup` (Curve?)

* Yeah, this one is a Curve (< ResourceObject < ParentObject < ModelObject)

>     * `Table:IndependentVariableList` (Use `ModelObjectList`?)

* ModelObjectList would work yes. But its behavior relating to clone/remove might be annoying (I believe it'll clone the underlying objects, regardless of whether they are ResourceObjects or not, so you might have to work around that... make sure to write extensive tests for that)

>     * `Table:IndependentVariable` (ParentObject?)

* I would say ResourceObject actually. It could technically be shared by several TableLookup objects. 
    * Eg: you have a CSV file (or you use the extensible fields instead of using an external file, doesn't matter conceptually) where the first column is the temperature (independent variable), and the second and third columns are the EIRFT and CAPFT values (dependent variables/output), then you would have two TableLookUp using the same Table:IndependentVariable object.

**Review by jmarrec — CHANGES_REQUESTED** [review]

File: resources/model/OpenStudio.idd [hunk 1]
```diff
@@ -21062,6 +21062,119 @@ OS:Table:MultiVariableLookup,
...
+       \key Temperature
+       \key VolumetricFlow
+       \key MassFlow
+       \key Power
+       \key Distance
+       \key Angle
+       \required-field
+  A6,  \field External File Name
+       \type alpha
+       \retaincase
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply resources/model/OpenStudio.idd]: 8ecd2254fd7311223941b834bf3c7dbbf5dd4637 and 8304c9d46265170b6302be4ea97c1dec07a6fd64

File: src/model/TableIndependentVariable.hpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    std::string unitType() const;
+
+    std::vector<TableLookup> tableLookups() const;
+
+    //@}
+    /** @name Setters */
+    //@{
+
+    bool setInterpolationMethod(std::string interpolationMethod);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/TableIndependentVariable.hpp]: https://github.com/NREL/OpenStudio/commit/8ecd2254fd7311223941b834bf3c7dbbf5dd4637 and https://github.com/NREL/OpenStudio/commit/8304c9d46265170b6302be4ea97c1dec07a6fd64

File: src/model/TableIndependentVariable.hpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    /** @name Constructors and Destructors */
+    //@{
+
+    explicit TableIndependentVariable(const Model& model);
+
+    virtual ~TableIndependentVariable() {}
+
+    //@}
+
+    static IddObjectType iddObjectType();
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/TableIndependentVariable.hpp:61]: https://github.com/NREL/OpenStudio/commit/8ecd2254fd7311223941b834bf3c7dbbf5dd4637 and https://github.com/NREL/OpenStudio/commit/8304c9d46265170b6302be4ea97c1dec07a6fd64

File: src/model/test/TableLookup_GTest.cpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+#include "../TableLookup.hpp"
+#include "../TableLookup_Impl.hpp"
+#include "../TableIndependentVariable.hpp"
+#include "../TableIndependentVariable_Impl.hpp"
+
+using namespace openstudio;
+using namespace openstudio::model;
+
+TEST_F(ModelFixture, TableLookup) {
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/test/TableLookup_GTest.cpp]: tests in a9d87a766dc9ab7adc0879731c498a0f80d8e79e (failing) plus fix in b4255acb4a535bbe74e0b19ec6b723b34c0ecef3

File: src/energyplus/ForwardTranslator/ForwardTranslateTableLookup.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+      eg.setDouble(Table_LookupExtensibleFields::OutputValue, outputValue);
+    }
+
+    // Table:IndependentVariableList
+    IdfObject tableIndependentVariableList(IddObjectType::Table_IndependentVariableList);
+    tableIndependentVariableList.setName(modelObject.nameString() + " Independent Variable List");
+    m_idfObjects.push_back(tableIndependentVariableList);
+
+    // IndependentVariableListName
+    tableLookup.setString(Table_LookupFields::IndependentVariableListName, tableIndependentVariableList.nameString());
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateTableLookup.cpp]: Done in 75a179ff0f8f6fdfa6929905b58b5660f4971dbd

File: src/energyplus/ForwardTranslator/ForwardTranslateTableLookup.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    // Table:IndependentVariable
+    for (auto& independentVariable : independentVariables) {
+      IdfObject tableIndependentVariable(IddObjectType::Table_IndependentVariable);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateTableLookup.cpp]: I redid the FT (and broke it in two) in f5d52a79cb97a18c847ad478111cb0e61591879f

File: src/model/TableLookup.hpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+    bool addOutputValue(double outputValue);
+
+    bool removeOutputValue(unsigned groupIndex);
+
+    void removeAllOutputValues();
+
+    std::vector<double> outputValues() const;
+
+    unsigned int numberofOutputValues() const;
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/TableLookup.hpp]: Done in dced8f0847bfc5717789b4864ee30cc2a5d0768c and eb66a23d786fe37a80c9e5e1771b3a7863868085

File: src/model/TableLookup.cpp [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+      return result;
+    }
+
+    IddObjectType TableLookup_Impl::iddObjectType() const {
+      return TableLookup::iddObjectType();
+    }
+
+    int TableLookup_Impl::numVariables() const {
+      return 1;
+    }
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/TableLookup.cpp]: cf 09380daf8743dfcf87baf005cf6374a4dcacf229

File: src/model/TableLookup.cpp [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+      : Curve_Impl(other, model, keepHandle) {
+      OS_ASSERT(other.iddObject().type() == TableLookup::iddObjectType());
+    }
+
+    TableLookup_Impl::TableLookup_Impl(const TableLookup_Impl& other, Model_Impl* model, bool keepHandle) : Curve_Impl(other, model, keepHandle) {}
+
+    const std::vector<std::string>& TableLookup_Impl::outputVariableNames() const {
+      static const std::vector<std::string> result;
+      return result;
+    }
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/TableLookup.cpp]: cf 09380daf8743dfcf87baf005cf6374a4dcacf229

**Comment by jmarrec** [comment]
@joseph-robertson I have made a significant amount of changes, could you review that? Still missing VersionTranslation rules and some more changes, cf TODO in https://github.com/NREL/OpenStudio/pull/4652#discussion_r959341446

File: src/energyplus/ForwardTranslator.cpp [hunk 10]
```diff
@@ -307,6 +307,131 @@ namespace energyplus {
       airWall.remove();
     }
 
+    // replace old TableMultiVariableLookup with TableLookup
+    for (TableMultiVariableLookup tableMulti : model.getModelObjects<TableMultiVariableLookup>()) {
+      OptionalString s;
+      OptionalDouble d;
+
+      TableLookup tableLookup(model);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator.cpp:315]: Added VT tests in f49be2690afe63955c3fa8e0f4abc2fd8630a542 and VT in f796e1cae355203352af4a38332dbdb7175b5b88

File: src/energyplus/ForwardTranslator.cpp [hunk 11]
```diff
@@ -307,6 +307,143 @@ namespace energyplus {
...
+
+      // Name
+      if (s = tableMulti.nameString()) {
+        tableLookup.setName(*s);
+      }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator.cpp:327]:
```suggestion
      // Name
      tableLookup.setName(tableMulti.nameString());
```

File: src/utilities/core/Deprecated.hpp [hunk 12]
```diff
@@ -57,4 +57,26 @@
...
+
+#  define DO_PRAGMA(X) _Pragma(#X)
+#  define DISABLE_WARNING_PUSH DO_PRAGMA(GCC diagnostic push)
+#  define DISABLE_WARNING_POP DO_PRAGMA(GCC diagnostic pop)
+#  define DISABLE_WARNING(warningName) DO_PRAGMA(GCC diagnostic ignored #warningName)
+
+#  define DISABLE_WARNING_DEPRECATED DISABLE_WARNING(-Wdeprecated-declarations)
+
+#endif
+// clang-format on
```
Comment (jmarrec) [inline src/utilities/core/Deprecated.hpp:80]: New macro for clarity/shortness. See it in example at https://godbolt.org/z/jEKzcedoK across 3 compilers

File: src/energyplus/ReverseTranslator.cpp [hunk 13]
```diff
@@ -909,6 +909,18 @@ namespace energyplus {
...
+      case openstudio::IddObjectType::Table_Lookup: {
+        modelObject = translateTableLookup(workspaceObject);
+        break;
+      }
+      case openstudio::IddObjectType::Table_IndependentVariableList: {
+        break;  // Done via translateTableLookup
+      }
+      case openstudio::IddObjectType::Table_IndependentVariable: {
+        break;  // Done via translateTableLookup
+      }
```
Comment (jmarrec) [inline src/energyplus/ReverseTranslator.cpp:922]: new RT

**Review by jmarrec — APPROVED** [review]
@joseph-robertson I added a ReverseTranslator. I opened a PR to https://github.com/NREL/OpenStudio-resources/pull/173 which caught a few bugs and I added an explicit test for sharing two TableIndependentVariables on two TableLookups.

File: src/utilities/core/Deprecated.hpp (hunk 12 above)
Comment (jmarrec) [reply src/utilities/core/Deprecated.hpp]: I removed it, swig was choking on it

File: src/energyplus/ForwardTranslator.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator.cpp]: Done everything in the latest commits

File: src/energyplus/ForwardTranslator/ForwardTranslateTableIndependentVariable.cpp [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    // Unit Type: String
+    idfObject.setString(Table_IndependentVariableFields::UnitType, modelObject.unitType());
+
+    // Value
+    for (const double& value : modelObject.values()) {
+      IdfExtensibleGroup eg = idfObject.pushExtensibleGroup();
+      eg.setDouble(Table_IndependentVariableExtensibleFields::Value, value);
+    }
```
[the remark this answers is not in the document]
Comment (joseph-robertson) [reply src/energyplus/ForwardTranslator/ForwardTranslateTableIndependentVariable.cpp:81]: @jmarrec Is this still a TODO?

**Pull request merged**